src/weapon/w_007.c

```diff
@@ -67,11 +239,100 @@ void EntityWeaponAttack(Entity* self) {
     self->rotPivotY = PLAYER.rotPivotY;
 }
 
-INCLUDE_ASM("weapon/nonmatchings/w_007", func_ptr_80170004);
-
-extern u8 D_35000_8017AC64[6][8];
+static void func_ptr_80170004(Entity* self) {
+    Entity* parent;
+    s32 temp_v1_3;
+    s32 temp_v1_4;
+    u16 a0;
+    u16 v0;
+
+    switch (self->step) {
+    case 0:
+        parent = self->ext.weapon.parent;
+        if (parent->entityId != 0) {
+            temp_v1_3 = (self->params >> 7);
+            self->facingLeft = parent->facingLeft;
+            a0 = D_35000_8017AC48[temp_v1_3];
+            v0 = D_35000_8017AC48[temp_v1_3 + 1];
+            if (self->facingLeft) {
+                a0 = -a0;
+            }
+            self->posX.i.hi += a0;
+            self->posY.i.hi += v0;
+
+            self->animSet = self->ext.weapon.parent->animSet;
+            self->anim = D_35000_8017AB44;
+            self->zPriority = 0xA8;
+            self->unk5A = self->ext.weapon.parent->unk5A;
+            self->palette = self->ext.weapon.parent->palette;
+
+            SetSpeedX(FIX(4.75));
+            temp_v1_4 = self->params & 0xFF00;
+            if ((temp_v1_4 == 0x300) || (temp_v1_4 == 0x600)) {
+                self->velocityY = FIX(2.75);
+                SetSpeedX(FIX(3.75));
+            }
+            self->drawFlags = FLAG_DRAW_ROTY | FLAG_DRAW_ROTX;
+            self->rotX = self->rotY = 0x10;
+            self->flags = FLAG_UNK_08000000 | FLAG_UNK_100000;
+            self->hitboxWidth = self->hitboxHeight = 9;
+            self->attack = 1;
+            self->attackElement = 0;
+            self->hitboxState = 4;
+            self->nFramesInvincibility = 4;
+            self->stunFrames = 4;
+            self->hitEffect = 1;
+            self->entityRoomIndex = 0;
+            g_api.func_80118894(self);
+            g_api.PlaySfx(SFX_UNK_6AD);
+            self->step++;
+            return;
+        }
+        DestroyEntity(self);
+        return;
+    case 1:
+        if (self->hitFlags) {
+            self->step = 3;
+        }
+        DecelerateX(FIX(7) / 32);
+        DecelerateY(FIX(7) / 32);
```

Comment:
This should be FIX(7.0 / 32), or FIX(28.0/128). The idea of FIX is to transform a fixed-point value, so we should include the integer and fractional portions together.